test/add/common/version_constraint_test.dart

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2020, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:test/test.dart';
+
+import '../../descriptor.dart' as d;
+import '../../test_pub.dart';
+
+void main() {
+  test('allows empty version constraint', () async {
+    await servePackages((builder) => builder.serve('foo', '1.2.3'));
```

Comment:
Maybe add a few more versions to the server here..

It's worth testing a few cases where:
 * There is older versions
 * There is multiple major versions
 * There is newer prereleases (major version)
 * There is newer prereleases (minor version)

Just cover a little bit more ground for the cases where the default logic picks a good version and version constraint.